### ISSUE TITLE
OpenAI CLIP -> OpenCLIP Conversion guide for H-14

### ISSUE THREAD

**Issue by trufty** [mlfoundations/open_clip#168]
Seeing that this repo holds the new SOTA CLIP model `H-14`, a lot of people are about to begin migrating to this pip package.

I think it would be a good idea to have a conversion guide to cover some of the differences between the two packages, to make the transition easier (It was a bit painful for me yesterday). 

So I'll list the differences and issues I ran into when converting my CLIP search app:
<br>

  **Lets start with the model loader:**
  OpenAI ->
  ```py
  model, preprocess = clip.load("./models/ViT-L-14-336px.pt", device=device)
  ```
  OpenClip ->
  ```py
  model, _, preprocess = open_clip.create_model_and_transforms('ViT-H-14', device=device, pretrained="laion2b_s32b_b79k")
  ```

- You can see `create_model_and_transforms` requires an additional `pretrained` method param to download the model.  If you don't include this, your code will happily chug away on an empty model XD.
- OpenClip also returns a size 3 tuple instead of 2 in OpenAI. 
- The type of `device` param is also different.  OpenClip requires `torch.device('cuda') or torch.device('cpu')` where OpenAI can take a string of `'cuda' or 'cpu'`

<br>
<br>

**Now we can move on to image preprocessing:**
OpenAI ->
```py
image = preprocess(Image.open(path)).to(device)
```
OpenClip ->
```py
image = preprocess(Image.open(path)).unsqueeze(0).to(device)
```

- `.unsqueeze(0)` is documented in the readme here, but I just overlooked it since my OpenAI implementation does not require it.  So make sure to include it in your code
- Don't forget to push it to the correct device with `.to(device)`

<br>
<br>

**Finally we can encode the image(s):**
OpenAI ->
```py
images = torch.stack(images)
image_features = model.encode_image(images).cpu().numpy()
```
OpenClip ->
```py
image_feature = model.encode_image(image).cpu().numpy()
```

- OpenAI allows a batch of images to be encoded at once.  OpenClip is one at a time (correct me if I'm wrong)
- Lastly the returned features shape is slightly different.  OpenAI l-14 `image_features[0]` looks like `(786,)` and OpenClip H-14 `image_feature` looks like `(1, 1024)`  (I think that was correct, it's been  a day since I saw the numbers)
    - This means `image_feature.tolist()` might unpack slightly differently with OpenLClip `image_feature.tolist()[0]` was my solution
    
<br>
<br>
    
Well, that's all I needed to convert to from OpenAI CLIP to OpenClip.  
I hope this helps someone else!

I appreciate all the effort put into this repo so far!

<br>
<br>

**Comment by rwightman** [comment]
@trufty one big difference to consider is that OpenAI interfaces are oriented towards inference, and this package covers both training (from scratch, from pretrained) and inference cases. We could at a specific creation API that's just for inference that forces a pretrained arg and only returns the inference transforms if you feel that'd help.... 
 * `create_model_for_inference`
 * or `create_model_from_pretrained` ?
@rom1504 any thoughts on that from the use cases you see?

The leading dimension is the batch dimension, torch stack stacks ads the leading dim (stacks into dim 0 by default), unsqueeze(0) makes the leading dim 1 for a single image case, these are the same.

`torch.stack([torch.randn(3, 224, 224)]).shape == torch.randn(3, 224, 224).unsqueeze(0).shape`

**Comment by grexzen** [comment]
Another thing to consider is that the outputs change with this. I tried it for scoring images, and outputs changed, so I had to revert, because that introduced another variable to my research.

Also, looking at the benchmark tables, the difference is negligible between H/14 and L/14.

**Comment by rom1504** [comment]
It's a different model yes.
And the improvement is quite significative on many evaluation datasets. On classification tasks but also particularly on retrieval https://laion.ai/blog/large-openclip/

**Comment by rwightman** [comment]
@trufty create_model_from_pretrained is merged, and device accepts either string or torch.device now
